### DISCUSSION TITLE
How can I remove and crop splats?

### DISCUSSION THREAD

**Discussion by gcampini** [mkkellogg/GaussianSplats3D#306] [Q&A] [answered]
Hi,

I have two main goals:

1. Hide splats outside an invisible box (cropping), so the user can only see the splats inside the box.
2. Remove the cropped splats completely from the viewer (maybe using splatmesh or something similar, not sure yet), and re-export the remaining splats (without the cropped ones) to a new .ply file or another compatible format.

I believe the first goal can be achieved by either instructing three.js to render only part of the scene or by dynamically modifying some splats before rendering via the lib.

As for the second goal, I’m honestly not sure how to proceed. I was considering using the parser output, but I’m unsure if that’s the best approach. I know there is a tool that does that (selecting and dynamically deleting splats) : https://playcanvas.com/supersplat/editor/

Any guidance or ideas would be greatly appreciated!
Thank you!

**Comment by mkkellogg** [comment, marked as answer]
For (1), you could conceivably pass 2 three-element vectors corresponding to the bounding box minimum and maximum as uniforms to the shader for `SplatMesh` and use those values to determine which splats to make visible. You could add the uniforms here: https://github.com/mkkellogg/GaussianSplats3D/blob/2d80511077fcb83ec5cb6da7336dc38e55d4a14b/src/splatmesh/SplatMaterial.js#L20
It might look like:
```
   uniform vec3 min;
   uniform vec3 max;
```
Then you can add logic to determine if the splat is in the bounding box by using the value of `splatCenter` before this line: https://github.com/mkkellogg/GaussianSplats3D/blob/2d80511077fcb83ec5cb6da7336dc38e55d4a14b/src/splatmesh/SplatMaterial.js#L156
The code would look something like:
```
 if (splatCenter.x < min.x || splatCenter.y < min.y || splatCenter.z < min.z || splatCenter.x > max.x || splatCenter.y > max.y || splatCenter.z > max.z) {
    gl_Position = vec4(0.0, 0.0, 2.0, 1.0);
    return;
}
```

For (2) I would say SuperSplat is your best option.

**Comment by jarodabel** [comment]
I have been working on a feature to crop the pointcloud, heres how I did it. 

https://github.com/mkkellogg/GaussianSplats3D/pull/321

In my web app I call this when the crop box handles are adjusted by the user 

```
      const splatMaterial =
        viewerRef.current?.splatsViewer?.splatMesh?.material;
      if (
        splatMaterial &&
        !Array.isArray(splatMaterial) &&
        splatMaterial instanceof THREE.ShaderMaterial
      ) {
        splatMaterial.uniforms.boxMin.value.set(
          frameEvent.frame.min.x,
          frameEvent.frame.min.y,
          frameEvent.frame.min.z
        );
        splatMaterial.uniforms.boxMax.value.set(
          frameEvent.frame.max.x,
          frameEvent.frame.max.y,
          frameEvent.frame.max.z
        );
      }
    },
```

![image](https://github.com/user-attachments/assets/b42fe335-ea69-4a78-ad22-bd84612349b1)
![image](https://github.com/user-attachments/assets/d78fd10b-9cb7-4397-920a-e7f317889573)
